### PR TITLE
Enables simple formatting

### PR DESCRIPTION
**Pull request by morris25**

### CODE REVIEW

File: src/styles.jl [hunk 1]
```diff
@@ -48,6 +48,27 @@ function styles_cell_xf_numFmtId(wb::Workbook, index::Int) :: Int
     return parse(Int, el["numFmtId"])
 end
 
+function styles_add_numFmt(wb::Workbook, format_code::AbstractString) :: Integer
```
Comment (omus) [inline src/styles.jl]: Docstring?

File: src/styles.jl [hunk 2]
```diff
@@ -48,6 +48,27 @@ function styles_cell_xf_numFmtId(wb::Workbook, index::Int) :: Int
...
+        stylesheet = findfirst(xroot, "/xpath:styleSheet", SPREADSHEET_NAMESPACE_XPATH_ARG)
+        numfmts = EzXML.addelement!(stylesheet, "numFmts")
+    else
+        numfmts = numfmts[1]
+    end
+
+    existing_numFmt_elements_count = EzXML.countelements(numfmts)
+
+    new_fmt = EzXML.addelement!(numfmts, "numFmt")
+    fmt_code = 164 + existing_numFmt_elements_count
```
Comment (omus) [inline src/styles.jl]: Magic numbers should have comments explaining them

File: src/write.jl [hunk 3]
```diff
@@ -192,8 +192,14 @@ function setdata!(ws::Worksheet, cell::Cell)
     nothing
 end
 
-function setdata!(ws::Worksheet, ref::CellRef, v::Union{Int, Float64})
-    cell = Cell(ref, "", "", string(v), "")
+function setdata!(ws::Worksheet, ref::CellRef, v::Union{Int, Float64}; formatid::Integer=-1)
+    if formatid > -1
```
Comment (omus) [inline src/write.jl]: Possibly easier to ready if you flipped the logic and did `formatid < 0`

File: src/styles.jl [hunk 4]
```diff
@@ -48,6 +48,27 @@ function styles_cell_xf_numFmtId(wb::Workbook, index::Int) :: Int
...
+    new_fmt["numFmtId"] = fmt_code
+    new_fmt["formatCode"] = format_code
+
+    return fmt_code
+end
```
Comment (omus) [inline src/styles.jl]: Maybe a Julia type should be made to represent this?

Comment (morris25) [reply src/styles.jl]: I will be posting a refactoring PR in the near future, so I'm holding off on adding a new type until then.

File: src/styles.jl [hunk 5]
```diff
@@ -48,6 +48,59 @@ function styles_cell_xf_numFmtId(wb::Workbook, index::Int) :: Int
...
+    # Any custom formats must have an id >= PREDEFINED_NUMFMT_COUNT
+    PREDEFINED_NUMFMT_COUNT = 164
+    fmt_code = existing_numFmt_elements_count + PREDEFINED_NUMFMT_COUNT
+    new_fmt["numFmtId"] = fmt_code
+    new_fmt["formatCode"] = format_code
+
+    return fmt_code
+end
+
+FontAttribute = Union{AbstractString, Pair{String, Pair{String, String}}}
```
Comment (felipenoris) [inline src/styles.jl]: Should be `const`?

**Comment by felipenoris** [comment]
In the same line as pointed out by @omus, if there's a `CellDataFormat` type, that should replace all occurences of `styleid` in the code.

Also, it's not good to `CellDataFormat` wrap the cell value.

If that is needed, a better design, in my opinion, would be to replace `CellValue` type, which is currently just an alias to raw Julia types, with something like

```julia
struct CellValue
   value::Union{ ... } # julia raw types in the current definition
   style::CellDataFormat # with style information
```

Then all `set_value` methods would accept `CellValue`s as input. And there would be another definition for `set_value` to accept raw julia datatypes with default formatting, which would be compatible with current code that uses `XLSX.jl`.

What do you think about this, @morris25 ?

**Comment by morris25** [comment]
Sounds fine to me

File: src/structs.jl [hunk 6]
```diff
@@ -42,10 +42,28 @@ struct EmptyCell <: AbstractCell
     ref::CellRef
 end
 
+abstract type AbstractCellDataFormat end
+struct EmptyCellDataFormat <: AbstractCellDataFormat end
+"""
+Keeps track of formatting information.
+"""
+struct CellDataFormat <: AbstractCellDataFormat
+    id::Integer
```
Comment (omus) [inline src/structs.jl]: Store as an `UInt` internally? How high do these formatting numbers go?

File: src/structs.jl [hunk 7]
```diff
@@ -42,10 +42,28 @@ struct EmptyCell <: AbstractCell
...
+
+    function CellDataFormat(id::Integer)
+        @assert id >= 0
+        new(id)
+    end
+end
+
+const CellValueTypes = Union{String, Missings.Missing, Float64, Int, Bool, Dates.Date, Dates.Time, Dates.DateTime}
```
Comment (omus) [inline src/structs.jl]: Should be singular

File: src/styles.jl [hunk 8]
```diff
@@ -48,6 +53,59 @@ function styles_cell_xf_numFmtId(wb::Workbook, index::Int) :: Int
...
+    else
+        numfmts = numfmts[1]
+    end
+
+    existing_numFmt_elements_count = EzXML.countelements(numfmts)
+    new_fmt = EzXML.addelement!(numfmts, "numFmt")
+
+    # The number of predefined number formats in XLSX
+    # Any custom formats must have an id >= PREDEFINED_NUMFMT_COUNT
+    PREDEFINED_NUMFMT_COUNT = 164
```
Comment (omus) [inline src/styles.jl]: Should probably be a `global const`

File: src/write.jl [hunk 9]
```diff
@@ -206,73 +200,59 @@ function xlsxescape(str::AbstractString)
...
-function setdata!(ws::Worksheet, ref::CellRef, str::AbstractString)
-    if isempty(str)
-        # no-op
-        return
+"""
+Returns the datatype and value for `val` to be inserted into `ws`.
+"""
+function xlsx_encode(ws::Worksheet, val::AbstractString)
+    if isempty(val)
+        return "", ""
```
Comment (omus) [inline src/write.jl]: I would put parenthesis around this so it is obvious you're returning a tuple

File: src/write.jl [hunk 10]
```diff
@@ -206,73 +200,59 @@ function xlsxescape(str::AbstractString)
...
-    value_str = b ? "1" : "0"
-    cell = Cell(ref, "b", "", value_str, "")
-    setdata!(ws, cell)
-end
-
-const DEFAULT_DATE_numFmtId = 14
-const DEFAULT_TIME_numFmtId = 20
-const DEFAULT_DATETIME_numFmtId = 22
+function get_style_index(ws::Worksheet, formatid::Integer)
+    @assert formatid >= 0
```
Comment (omus) [inline src/write.jl]: Should you just make a `CellDataFormat` here?

File: src/write.jl [hunk 11]
```diff
@@ -206,73 +200,59 @@ function xlsxescape(str::AbstractString)
...
-        time_style_index = styles_add_cell_xf(wb, Dict("applyNumberFormat"=>"1", "borderId"=>"0", "fillId"=>"0", "fontId"=>"0", "numFmtId"=>"20", "xfId"=>"0"))
-    end
-
-    value_str = string(time_to_excel_value(t))
-    cell = Cell(ref, "", string(time_style_index), value_str, "")
-    setdata!(ws, cell)
-end
+const DEFAULT_DATE_numFmtId = 14
+const DEFAULT_TIME_numFmtId = 20
+const DEFAULT_DATETIME_numFmtId = 22
```
Comment (omus) [inline src/write.jl]: Maybe would be good to have a comment which shows what this format looks like?

File: src/structs.jl [hunk 12]
```diff
@@ -42,10 +42,28 @@ struct EmptyCell <: AbstractCell
...
 """
 CellValue is a Julia type of a value read from a Spreadsheet.
 """
-const CellValue = Union{String, Missings.Missing, Float64, Int, Bool, Dates.Date, Dates.Time, Dates.DateTime}
+struct CellValue
+    value::CellValueTypes
+    styleid::AbstractCellDataFormat
+end
```
Comment (omus) [inline src/structs.jl]: Should this have another constructor like this:
```julia
CellValue(value::CellValueTypes) = ...
```
which just uses the default style ID?

File: src/write.jl (hunk 10 above)
Comment (morris25) [reply src/write.jl]: No, its a numFmtId. I'll probably replace it with a type but I want to leave that for when I add a bunch of other fomatting types and do a single wave of deprecations.

**Comment by felipenoris** [comment]
@morris25 , please let me know when this is ready, ok?

**Comment by morris25** [comment]
Sorry I didn't respond sooner. I was at a conference. So long as you're ok with the code, this set of changes should be good to go. I'll be adding some other features in subsequent PRs.

**Pull request merged**